libs/actions/common/PythonRunner/pythonrunner.py

```diff
@@ -0,0 +1,180 @@
+
+    @classmethod
+    def _get_module_names_from_script(cls, script_path: str) -> set:
+        """
+        Get the names of the modules that are imported in the script.
+
+        :param script_path: [str] The path to the script.
+        :return: [set] The set of module names.
+        """
+        with open(script_path, "r") as file:
+            lines = file.readlines()
+        imports = [
+            line.strip()
+            for line in lines
+            if line.strip().startswith(("import", "from"))
+        ]
+        logger.info(f"Found import lines: {imports}")
+        packages = set()
+        for line in imports:
+            parts = line.split()
+            module_name = parts[1].split(".")[0]
+            if (
+                module_name not in sys.builtin_module_names
+                and module_name not in sys.stdlib_module_names
+            ):
+                packages.add(module_name)
+        return packages
+
+    @classmethod
+    def _get_venv_paths(cls, venv_dir: str) -> tuple[str, str]:
+        """
+        Get the paths to the python and pip executables in the virtual environment.
+
+        :param venv_dir: [str] The path to the virtual environment directory.
+        :return: [tuple] The paths to the python and pip executables.
+        """
+
+        bin_folder = "Scripts" if sys.platform == "win32" else "bin"
+        venv_bin = os.path.join(venv_dir, bin_folder)
+        venv_python = os.path.join(venv_bin, "python")
+        venv_pip = os.path.join(venv_bin, "pip")
+        return venv_python, venv_pip
+
+    @classmethod
+    def _install_packages(cls, venv_pip: str, packages: list) -> None:
+        """
+        Install the provided packages in the virtual environment.
+
+        :param venv_pip: [str] The path to the pip executable in the virtual environment.
+        :param packages: [list] The list of package names to install.
+        """
+        if packages:
+            command = [venv_pip, "install"] + packages
+            try:
+                subprocess.run(command, check=True)
+            except subprocess.CalledProcessError as e:
+                logger.error(f"Error installing packages: {e}")
+
+    @classmethod
+    def _check_and_install_virtualenv(cls) -> None:
+        """
+        Check if virtualenv is installed, if not, install it.
+        """
+        try:
+            import virtualenv
+
+            logger.info("virtualenv is already installed.")
+        except ImportError:
+            logger.info("virtualenv is not installed, installing now...")
+            subprocess.check_call(
+                [sys.executable, "-m", "pip", "install", "virtualenv"]
+            )
+
+    @classmethod
+    def check_python_file_or_folder(cls, path: str, venv_pip: str) -> str:
+        if os.path.isdir(path):
+            requirements_path = os.path.join(path, "requirements.txt")
+            main_script_path = os.path.join(path, "main.py")
+            if os.path.exists(requirements_path):
```

Comment:
@pisces860225, please use early return to eliminate the extra indentations.

Reply:
I have already performed a simple refactoring and further decomposed the logic, splitting it into two functions to accomplish this task, reducing the "god object" phenomenon.